dll/win32/iphlpapi/icmp.c

```diff
@@ -1,1129 +1,575 @@
-                    TRACE("Got rtt = %u\n", reply->RoundTripTime);
-                }
-            }
-
-            memmove(buf, eol + 1, pos - (eol + 1 - buf));
-            pos -= (eol + 1 - buf);
+            CopyMemory(
+                (PBYTE)InputBuffer + sizeof(ICMP_ECHO_REQUEST),
+                RequestOptions->OptionsData,
+                Request->OptionsSize);
         }
     }
-    while (res > 0);
-    close(pipe_out[0]);
 
-    /* reap the child process */
-    do
+    if (RequestSize > 0)
     {
-        wpid = waitpid(pid, &status, 0);
+        Request->DataSize = RequestSize;
+        CopyMemory((PBYTE)InputBuffer + Request->DataOffset, RequestData, RequestSize);
     }
-    while (wpid < 0 && errno == EINTR);
 
-    /* fill out remaining struct fields */
-    if (wpid >= 0 && WIFEXITED(status) && WEXITSTATUS(status) == 0)
+    if (Event == NULL && ApcRoutine == NULL)
     {
-        if (ReplySize < RequestSize + sizeof(*reply))
-        {
-            reply->Status   = IP_BUF_TOO_SMALL;
-            reply->DataSize = 0;
-            reply->Data     = NULL;
-        }
-        else
-        {
-            reply->Status   = 0;
-            reply->DataSize = RequestSize;
-            reply->Data     = (char *)reply + sizeof(*reply);
-            memcpy(reply->Data, RequestData, RequestSize);
-        }
-        return 1;
+        hEvent = CreateEventW(NULL, FALSE, FALSE, NULL);
     }
-
-    SetLastError(IP_REQ_TIMED_OUT);
-    return 0;
-#else
-    ERR("no fork support on this platform\n");
-    SetLastError(ERROR_NOT_SUPPORTED);
-    return 0;
-#endif
-}
-#else // __REACTOS__
-BOOL
-GetIPv4ByIndex(
-    _In_  DWORD              Index,
-    _Out_ IPAddr *           Address
-)
-{
-    PMIB_IPADDRTABLE pIpAddrTable;
-    ULONG dwSize = 0;
-    BOOL result = FALSE;
-
-    if (GetIpAddrTable(NULL, &dwSize, FALSE) != ERROR_INSUFFICIENT_BUFFER)
+    else
+    {
+        hEvent = Event;
+    }
+
+    Status = NtDeviceIoControlFile(
+        IcmpHandle,
+        hEvent,
+        ApcRoutine,
+        ApcContext,
+        &IoStatusBlock,
+        IOCTL_ICMP_ECHO_REQUEST,
+        InputBuffer,
+        ReplySize,
+        ReplyBuffer,
+        ReplySize);         // TODO: Determine how Windows calculates OutputBufferLength.
+
+    // If called asynchronously, return for the caller to handle.
+    if (Event != NULL || ApcRoutine != NULL)
     {
-        return result;
+        SetLastError(RtlNtStatusToDosError(Status));
+        HeapFree(GetProcessHeap(), 0, InputBuffer);
+        return 0;
     }
-    pIpAddrTable = HeapAlloc(GetProcessHeap(), 0, dwSize);
 
-    if (GetIpAddrTable(pIpAddrTable, &dwSize, FALSE) == NO_ERROR)
+    // Otherwise handle it like IcmpSendEcho.
+
+    if (Status == STATUS_PENDING)
     {
-        INT i;
+        Status = NtWaitForSingleObject(hEvent, FALSE, NULL);
 
-        for (i = 0; i < (*pIpAddrTable).dwNumEntries; i++)
+        if (NT_SUCCESS(Status))
```

Comment:
`NtWaitForSingleObject` never returns failure unless you pass an invalid or timeout pointer.

Reply:
I copied this pattern from other code calling `NtDeviceIoControlFile()`.